File: WdtcCore/src/main/java/org/wdt/wdtc/core/download/quilt/QuiltVersionList.java
package org.wdt.wdtc.core.download.quilt;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.annotations.SerializedName;
import org.wdt.utils.gson.JsonArrayUtils;
import org.wdt.utils.gson.JsonObjectUtils;
import org.wdt.wdtc.core.download.infterface.VersionJsonObjectInterface;
import org.wdt.wdtc.core.download.infterface.VersionListInterface;
import org.wdt.wdtc.core.game.Launcher;
import org.wdt.wdtc.core.utils.URLUtils;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class QuiltVersionList implements VersionListInterface {
  private static final String QuiltVersionListUrl = "https://meta.quiltmc.org//v3/versions/loader/%s";
  private final Launcher launcher;

  public QuiltVersionList(Launcher launcher) {
    this.launcher = launcher;
  }

  @Override
  public List<VersionJsonObjectInterface> getVersionList() throws IOException {
    List<VersionJsonObjectInterface> List = new ArrayList<>();
    JsonArray VersionArray = JsonArrayUtils.parseJsonArray(URLUtils.getURLToString(String.format(QuiltVersionListUrl, launcher.getVersionNumber())));
    for (int i = 0; i < VersionArray.size(); i++) {
      JsonObject VersionObject = VersionArray.get(i).getAsJsonObject();
      List.add(JsonObjectUtils.parseObject(VersionObject.getAsJsonObject("loader"), QuiltVersionJsonObjectImpl.class));
    }
    return List;
  }

  public static class QuiltVersionJsonObjectImpl implements VersionJsonObjectInterface {
    @SerializedName("version")
    private String versionNumber;
    @SerializedName("build")
    private int buildNumber;

    @Override
    public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;
      QuiltVersionJsonObjectImpl that = (QuiltVersionJsonObjectImpl) o;
      return buildNumber == that.buildNumber && Objects.equals(versionNumber, that.versionNumber);
    }

    @Override
    public int hashCode() {
      return Objects.hash(versionNumber, buildNumber);
    }

    @Override
    public String getVersionNumber() {
      return versionNumber;
    }

  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/download/quilt/QuiltVersionList.kt
package org.wdt.wdtc.core.download.quilt

import com.google.gson.annotations.SerializedName
import org.wdt.utils.gson.getJsonObject
import org.wdt.utils.gson.parseJsonArray
import org.wdt.utils.gson.parseObject
import org.wdt.utils.io.toStrings
import org.wdt.wdtc.core.download.infterface.VersionListInterface
import org.wdt.wdtc.core.download.infterface.VersionsJsonObjectInterface
import org.wdt.wdtc.core.game.*
import org.wdt.wdtc.core.manger.officialDownloadSource
import org.wdt.wdtc.core.utils.toURL
import java.io.IOException
import java.util.*

class QuiltVersionList(private val version: Version) : VersionListInterface {
	private val quiltquVersionListUrl = "${officialDownloadSource.quiltMetaUrl}v3/versions/loader/%s"
	
	@get:Throws(IOException::class)
	override val versionList: LinkedList<VersionsJsonObjectInterface>
		get() {
			return LinkedList<VersionsJsonObjectInterface>().apply {
				quiltquVersionListUrl.format(version.versionNumber).toURL().toStrings().parseJsonArray().forEach {
					add(it.asJsonObject.getJsonObject("loader").parseObject<QuiltVersionsJsonObjectImpl>())
				}
			}
		}
	
	class QuiltVersionsJsonObjectImpl(
		@field:SerializedName("version")
		override val versionNumber: String
	) : VersionsJsonObjectInterface {
		
		@SerializedName("build")
		var buildNumber = 0
		override fun equals(other: Any?): Boolean {
			if (this === other) return true
			if (other == null || javaClass != other.javaClass) return false
			val that = other as QuiltVersionsJsonObjectImpl
			return buildNumber == that.buildNumber && versionNumber == that.versionNumber
		}
		
		override fun hashCode(): Int {
			return Objects.hash(versionNumber, buildNumber)
		}
		
		
	}
	
}
